### PR TITLE
Patch/20170504

### PR DESCRIPTION
**Pull request by kazuyaujihara**
Miscellaneous patches.

### CODE REVIEW

**Comment by egonw** [comment]
Dear Kazuya, first, welcome!

Oh, wow! Wonderful! There are some patches that should be applied with high priority, but it also has patches I need to look more close too.

@kazuyaujihara, a general would be for me to ask if you could indicate (where needed, and some patches are evident), *what* is being fixed? For example, what problem is fixed by commit https://github.com/cdk/cdk/pull/317/commits/c0046791c2e889c8a47d898cf90d593d92b3387b . Also, we have as preference to have unit tests that first demonstrate the problem, e.g. because the existing code calculates the wrong answer. That makes it easier for the reviewer to verify the code, as we do no always manage to get the original author to look at the patch.

@johnmay, @kazuyaujihara, may I suggest we cherry-pick the easy ones (e.g. the one about the readers), and then we can review the remaining patches in more detail?

E.g. I think these patches are good to go:

- https://github.com/cdk/cdk/pull/317/commits/dc3ee7636e61c73d90e109dcd683a468f41fbc5f
- https://github.com/cdk/cdk/pull/317/commits/50a6127bbf22e4d9da8f937b4474480f37dcf267
- https://github.com/cdk/cdk/pull/317/commits/be89e6667990c856a73066a90f658d3b0ba5edd6
- https://github.com/cdk/cdk/pull/317/commits/59054f13ec0c38923c814c9b38c26e9c29c1725f
- https://github.com/cdk/cdk/pull/317/commits/e5df9ff5189eb5dac025efb70c793997a6504a55

And these ones probably too:

- https://github.com/cdk/cdk/pull/317/commits/bd675612fffb0ba7a37b67229271c5eaddde3016
- https://github.com/cdk/cdk/pull/317/commits/d6ca462ec02b2da45d39ba0dd2751afdc93ea744

@johnmay, what do you think? Shall I cherry-pick these, and so that this PR can then be rebased?

**Comment by kazuyaujihara** [comment]
Dear Egon, thank you for accepting me.

Because these patches are logically found on my way to port CDK to C#, I do
not have any unit tests yet. So I will explain each of them tentatively.

f335355
<https://github.com/cdk/cdk/pull/317/commits/f33535550e75ad5e4ce7c10fda6e0328d4914c2d>
---
The file name is 2,5-diMe-furan but the content is 2,5-diMe-pyrrole. Either
of two should be corrected.
fe2e31f
<https://github.com/cdk/cdk/pull/317/commits/fe2e31fc9df3bd7ffd44cc44c68244983c3d50b5>
---
There are the resemble codes in Rearrangement...Reaction classes, but only
RearrangementRadicalReaction class is different. Need a unit test.
5de731a
<https://github.com/cdk/cdk/pull/317/commits/5de731a4c42f82b2d7e92084ccd620b07baa0b54>
--- Because IsotopePatternRule.setParameters(Object[] params) accepts
List<double[]>
as params[0], getParameters()[0] is expected to return List<double[]>.
d6ca462
<https://github.com/cdk/cdk/pull/317/commits/d6ca462ec02b2da45d39ba0dd2751afdc93ea744>,
0c0bb18
<https://github.com/cdk/cdk/pull/317/commits/0c0bb1876ef0892d8927a76c0aa41a7983410643>,
a22d04f
<https://github.com/cdk/cdk/pull/317/commits/a22d04fa3b029f117e8410f98459d3f44e316873>,
a3af9ce
<https://github.com/cdk/cdk/pull/317/commits/a3af9ce67000329313f238b984b37f7569187b84>
--- Apparently,
they are typo or typical mistakes. Need unit tests.
5687f86
<https://github.com/cdk/cdk/pull/317/commits/5687f868e1000397977b6267f87f79f200f33937>
--- This line never returns null and null is returned at last line on the
proper condition.
a66b51f
<https://github.com/cdk/cdk/pull/317/commits/a66b51f212e89bc14950972d792ef21b60f32c9b>
--- I believe logical And/Or should be used in this context.
d759659
<https://github.com/cdk/cdk/pull/317/commits/d7596597815160d65f8df60e619d9affe431ca78>
--- I don't know there are other meaning or not.
c004679
<https://github.com/cdk/cdk/pull/317/commits/c0046791c2e889c8a47d898cf90d593d92b3387b>
--- I expect dblEqual(double v1, double v2) function is designed as nearly
equal function. But it does not work correctly when v1 or v2 are negative.
All current unit tests pass v1 == v2 to dblEqual. Need a unit test?
52a2279
<https://github.com/cdk/cdk/pull/317/commits/52a2279085b2bd899edc5aabcfcc4a60c01326b2>
--- My image of NoSuchElementException class is not used on this context.









2017-05-05 15:30 GMT+09:00 Egon Willighagen <notifications@github.com>:

File: base/core/src/main/java/org/openscience/cdk/atomtype/CDKAtomTypeMatcher.java [hunk 1]
```diff
@@ -1016,7 +1017,7 @@ private boolean isSingleHeteroAtom(IAtom atom, IAtomContainer container) {
 
     }
 
-    private boolean isRingAtom(IAtom atom, IAtomContainer atomContainer, RingSearch searcher) {
+    private boolean isRingAtom(IAtom atom, IAtomContainer atomContainer, RingSearch searcher) throws NoSuchAtomException {
```
Comment (johnmay) [inline base/core/src/main/java/org/openscience/cdk/atomtype/CDKAtomTypeMatcher.java]: Nope!!!! I'm actually about the change the API to try and make NoSuchAtom an unchecked exception but please watch out for breaking the API.

File: tool/forcefield/src/main/java/org/openscience/cdk/modeling/builder3d/MMFF94BasedAtomTypePattern.java [hunk 2]
```diff
@@ -163,7 +163,7 @@ private void createPattern() {
         //cdk.bug 3515122 fixed
         atomTypePatterns.add(Pattern.compile("N-[1-3];[CHN]{1,3}.{1}+[A-Z]{0,3}+[,]?+=O[CNXO].*+"));
         //NC=0 amid
-        atomTypePatterns.add(Pattern.compile("N-1-2];[CH]{1}+=S[(].*+"));
+        atomTypePatterns.add(Pattern.compile("N-[1-2];[CH]{1}+=S[(].*+"));
```
Comment (johnmay) [inline tool/forcefield/src/main/java/org/openscience/cdk/modeling/builder3d/MMFF94BasedAtomTypePattern.java]: Maybe... but this class should be deprecated and has been replaced by [Mmff](http://cdk.github.io/cdk/2.0/docs/api/index.html). If it's not deprecated I'll get on it.

**Comment by johnmay** [comment]
Patches look great thanks for tracking them done. Yes please cherry pick @egonw - am okay with the ones you've selected so far. Others are also okay except for the NoSuchElementException (see comment), the cdk-reaction one's are tricky.

**Comment by johnmay** [comment]
Ignore that approve... I though I could do it per commit.

**Comment by johnmay** [comment]
@egonw easier to list the ones that need more looking into:

fe2e31f - cdk-reaction need in depth look
a22d04f - cdk-reaction need in depth look
5de731a  - okay to my eye by @egonw you'll no more.
c004679 - need to look closer, maybe check with Alex Clark
52a2279 - API Change, am considering fixing this ATM on my local branch (NoSuchAtom should be unchecked)


Others that are okay and why
 - f335355 fine, see filename
 - 0c0bb18 fine, this was my bad :-)
 - 5687f86 fine, Map<IReaction,Integer>
 - a66b51f fine, no change in meaning but more correct for logical conditions
 - a3af9ce fine, my bad again
 - e5df9ff fine - duplicate code
 - d759659 fine - duplicate code
 - 59054f1 fine

**Comment by johnmay** [comment]
Thanks again.

I had a careful look at the ones I was unsure on and agree the patches are correct and merged all but the last one into master. I'll redo the  NoSuchAtomException once #318 is applied (or feel free to do it yourself). Out of interest how did you find these errors, lots of them seem very specific?

**Comment by johnmay** [comment]
Well thanks very nice finds, just reviewing the other pull request now. Did you use IKVM for C# or port the code? Might be something we can host.

**Comment by kazuyaujihara** [comment]
All codes are written in C#. The repository is the following.
https://github.com/kazuyaujihara/NCDK
I am happy to host it on the cdk.

**Pull request closed without merging**